presto-main-base/src/main/java/com/facebook/presto/operator/scalar/MathFunctions.java

```diff
@@ -1685,6 +1686,68 @@ public static double arrayL2SquaredDouble(
         return sum;
     }
 
+    @Description("Dot Product distance between the given identical sized vectors represented as DOUBLE arrays")
+    @ScalarFunction("dot_product")
+    @SqlNullable
+    @SqlType(StandardTypes.DOUBLE)
+    public static Double arrayDotProduct(
+            @SqlType("array(double)") Block leftArray,
+            @SqlType("array(double)") Block rightArray)
+    {
+        checkCondition(
+                leftArray.getPositionCount() == rightArray.getPositionCount(),
+                INVALID_FUNCTION_ARGUMENT,
+                "Both array arguments must have identical sizes");
+        double result = 0.0;
+
+        if (leftArray.mayHaveNull() || rightArray.mayHaveNull()) {
+            boolean hasNull = false;
+            for (int i = 0; i < leftArray.getPositionCount(); i++) {
+                hasNull = hasNull || leftArray.isNull(i) || rightArray.isNull(i);
+            }
+            if (hasNull) {
+                return null;
+            }
+        }
+
+        for (int i = 0; i < leftArray.getPositionCount(); i++) {
+            result += DOUBLE.getDouble(leftArray, i) * DOUBLE.getDouble(rightArray, i);
+        }
+
+        return result;
+    }
+
+    @Description("Dot Product distance between the given identical sized vectors represented as REAL arrays")
+    @ScalarFunction("dot_product")
+    @SqlNullable
+    @SqlType(StandardTypes.REAL)
+    public static Long arrayDotProductReal(
+            @SqlType("array(real)") Block leftArray,
+            @SqlType("array(real)") Block rightArray)
+    {
+        checkCondition(
+                leftArray.getPositionCount() == rightArray.getPositionCount(),
+                INVALID_FUNCTION_ARGUMENT,
+                "Both array arguments must have identical sizes");
+        float dotProduct = 0.0F;
+
+        if (leftArray.mayHaveNull() || rightArray.mayHaveNull()) {
+            boolean hasNull = false;
+            for (int i = 0; i < leftArray.getPositionCount(); i++) {
+                hasNull = hasNull || leftArray.isNull(i) || rightArray.isNull(i);
+            }
+            if (hasNull) {
+                return null;
+            }
+        }
+
+        for (int i = 0; i < leftArray.getPositionCount(); i++) {
+            dotProduct += Float.intBitsToFloat((int) REAL.getLong(leftArray, i)) * Float.intBitsToFloat((int) REAL.getLong(rightArray, i));
+        }
+
+        return ((long) Float.floatToRawIntBits(dotProduct));
+    }
+
```

Comment:
I thought we want to error out on NULL? if so, just check_argument !mayhavenulls upfront for both sides